### DISCUSSION TITLE
MongoClient partecipate to transaction

### DISCUSSION THREAD

**Discussion by luca-bassoricci** [quarkusio/quarkus#38926] [Q&A]
I'm currently using mongodb support (plain one, no Panache): is there any way to let MongoClient partecipate to current transaction?
I wrote a `@Decorator` around `MongoClient` but it is removed during build; made it @unremovable doesn't resolve the problem because the decorator in never called.

Any idea?
Regards Luca

[a comment by an automated account is not shown]

**Reply by loicmathieu** [reply]
Hi,
Two solutions here:
- You wrote your own transaction support and use the MongoDB client directly. It's not that simple but it can be done. I don't know what is an `@Decorator` so I don't know why it is not called, a reproducer project may help.
- You use MongoDB with Panache as it comes with transaction support and you can still use the underlying MongoDB collection so you can still do whatever you want.

I strongly suggest the second option but I may be biased as I wrote the extension ;)

**Reply by luca-bassoricci** [reply]
Thanks for your quick response.
About decorator, [this reproducer](https://github.com/luca-bassoricci/mongo-client-decorator-issue) contains a simple decorator around a MongoClient and a test; this test fails because decorator is never called (it is removed during build).

- Transaction support will require too much time and this solution has been discarded
- MongoDB with Panache give me `Panache::mongoCollection()` but this method doesn't partecipate to active transaction; it's necessary a call to `Panache::getSession()` and/or manage transaction manually and I prefer to stick with `@Transactional` support.

Currently I'm using quarkus-morphia.
When I tried to decorate the synthetic Morphia's `Datastore` created from extension because I want to add transactional support and I faced the same problem I had decorating the `MongoClient`. 
Could be a problem related to synthentic beans?

**Reply by loicmathieu** [reply]
You can open an issue for the decorator being removed, this may be because MongoClient is a synthetic bean or not, this is a question for a CDI guru ;)

In MongoDB, to participate in a transaction you must pass the ClientSession on each call to the driver.
This is why MongoDB with Panache is convenient, we do this for you ;)

So if you don't want to use MongoDB with Panache, you have to do the hard work by yourself.

**Comment by luca-bassoricci** [comment]
> You can open an issue for the decorator being removed, this may be because MongoClient is a synthetic bean or not, this is a question for a CDI guru ;)

I'll do.

> In MongoDB, to participate in a transaction you must pass the ClientSession on each call to the driver.
> This is why MongoDB with Panache is convenient, we do this for you ;)

I have a small question about this use case using panache:
```
@Transactional
public void doStuff()
{
  MyMongoEntity.mongoCollection().aggregate(...);
  MyMongoEntity.mongoCollection().findOneAndDelete(...);
}
```
in this case Panache doesn't pass to `MongoCollection::aggregate()` any client session: should it do it?

In https://github.com/quarkusio/quarkus/blob/b36adcbdfd309b1530635eecd39d7255c7c59125/extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/binder/CommonQueryBinder.java#L66 instead `value` should be encoded using the registered `Codec` instead of `value.toString()`?

**Reply by loicmathieu** [reply]
> in this case Panache doesn't pass to MongoCollection::aggregate() any client session: should it do it?

If you use directly the MongoCollection, you're on your own so you must pass the ClientSession on each method call on the collection.

> instead value should be encoded using the registered Codec instead of value.toString()?

Maybe, if you have a use case with a reproducer you can open an issue so we can discuss this.
